README.rst
Introduction
============

This Zope add-on product provides for easy configuration of ZServer Views.

It was authored by Leonardo Rochael Almeida, and sponsored by OpenMultimedia.

A ZServer View is a (supposedly) small and fast WSGI_ application that runs
directly from within the ZServer thread (a.k.a. medusa thread).

This means that a ZServer View will run even if all Zope worker-threads are
busy handling other requests. It also means that if a ZServer View is not fast,
then it is blocking the ZServer from handling incoming requests for all worker
threads, so be mindful when using this product.

ZServer Views are mostly useful for system monitoring tools, like
`DeadlockDebugger` or Zope resource monitoring.

Installation
============

Add `Products.ZServerViews` to the list of eggs of the part that defines your
Zope instance.

Configuration
=============

Add a `product-config` section for `zserver-views` to your Zope
configuration (e.g. the `zope.conf` file or a `zope-conf-additional` parameter
in your buildout configuration).

Each entry in this section should contain an identifier (must be unique within
the section but are currently unused [1]_), a URL path and the dotted
name of the view which will be served from that URL path, all separated by
whitespace. For example::

  <product-config zserver-views>
      my-package /my/view my.package.mymodule.my_view
      thread-id /thr-id Products.ZServerViews.tests.common.current_thread_id_zserver_view
  </environment>

This means that the `http://yourserver/my/view` URL will run the WSGI app at
`my.package.mymodule.my_view` and the url at `http://yourserver/thr-id` will
run the `Products.ZServerViews.tests.common.current_thread_id_zserver_view`
WSGI view.

Query strings, if present, will be passed to the ZServer Views, but sub-paths
will not. I.e., `http://yourserver/my/view?foo=bar` will run
`my.package.mymodule.my_view`, but `http://yourserver/thrid/fred` will not run
the `Products.ZServerViews.tests.common.current_thread_id_zserver_view` view,
but will be handled by Zope directly.

This behaviour is intentional, and reflects the fact that ZServer View location
is decided by a single dictionary lookup, instead of multiple regex or
substring matching.

.. [1] The identifiers in `product-config` keys cannot be composed
   of many valid URL characters, and are forcibly lower-cased, so they're
   useless for either the URL or the dotted name of the view.

Writing ZServer Views
=====================

ZServer Views are mini WSGI apps, which mean they take an `env` and a
`start_response` parameter, and return the bytes that must be
sent to the browser. The query string, if any, will be passed into
`env['QUERY_STRING']`.

For HTTP/1.1 compatibility they must either pass a "Content-Lenght" header into
`start_response()` or a "Transfer-Encoding=chunked" header and chunk the
content. If neither is done, the response will fall back to `HTTP/1.0`
automatically, and the connection will be forced close.

To ease the creation of these mini apps, a decorator exists
(`Products.ZServerViews.base import TextView`) that can wrap
a function receiving only the `env` parameter and returns a single `unicode`
string. This decorator will take care of all the WSGI conversation around that
response.

Check `Products.ZServerViews.tests.common.current_thread_id_zserver_view` for
an example.

.. _WSGI: http://wsgi.org/


docs/HISTORY.txt
Changelog
=========

0.2.1 (unreleased)
------------------

- Nothing changed yet.


0.2.0 (2013-01-24)
------------------

- Make the message of ViewError a mandatory parameter. It can be empty but lots
  of HTTP error codes require a body in the response, so an empty body is no
  longer the default [leorochael]

- Add a modicum of Python 2.4/Zope 2.10 compatibility (Exceptions are not
  new-style classes) [leorochael]

0.1.1 (2013-01-18)
------------------

- Include .rst files in the manifest [leorochael]


0.1.0 (2013-01-18)
------------------

- Initial release


Products/ZServerViews/tests/common.py
##############################################################################
#
# Copyright (c) 2013 Zope Foundation and Contributors.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE
#
##############################################################################

import thread
from Products.ZServerViews.base import TextView

def current_thread_id():
    return str(thread.get_ident())

@TextView
def current_thread_id_zserver_view(environment):
    return current_thread_id()

class App(object):
    "A publishable app"
    # docstring above necessary for this object to be traversable

    def __bobo_traverse__(self, REQUEST, name):
        return self

    def __call__(self, REQUEST):
        return current_thread_id()

# Enable this module to be published by ZServer
bobo_application = App()

import logging
LOG = logging.getLogger('ZServerPublisher')

# XXX: Horrible, horrible patch because the experimental implementation of
# wsgi on Zope broke regular module publishing for anything other than a module
# called "Zope2"
def patch_ZServerPublisher():

    # define the patched function here to keep the same indentation
    def ZServerPublisher__init__(self, accept):
        from sys import exc_info
        from ZPublisher import publish_module
        from ZPublisher.WSGIPublisher import publish_module as publish_wsgi
        while 1:
          try:
            name, a, b=accept()
            if name == "Zope2WSGI":
                try:
                    res = publish_wsgi(a, b)
                    for r in res:
                        a['wsgi.output'].write(r)
                finally:
                    # TODO: Support keeping connections open.
                    a['wsgi.output']._close = 1
                    a['wsgi.output'].close()
            else: # originally "if name == "Zope2"
                try:
                    publish_module(
                        name,
                        request=a,
                        response=b)
                finally:
                    b._finish()
                    a=b=None

          except:
            LOG.error('exception caught', exc_info=True)
    from ZServer.PubCore.ZServerPublisher import ZServerPublisher
    _original = ZServerPublisher.__init__
    ZServerPublisher__init__._original = _original
    ZServerPublisher__init__.__name__ = _original.__name__
    ZServerPublisher.__init__ = ZServerPublisher__init__

def unpatch_ZServerPublisher():
    from ZServer.PubCore.ZServerPublisher import ZServerPublisher
    ZServerPublisher.__init__ = ZServerPublisher.__init__._original


Products/ZServerViews/tests/__init__.py
##############################################################################
#
# Copyright (c) 2013 Zope Foundation and Contributors.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE
#
##############################################################################


Products/ZServerViews/tests/test_configuration.py
##############################################################################
#
# Copyright (c) 2013 Zope Foundation and Contributors.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE
#
##############################################################################

import cStringIO

from tempfile import mkdtemp
from shutil import rmtree

import ZConfig

from ZConfig.components.logger.tests import test_logger

# NOTE: Large sections of this file were pulled from
# Zope2.Startup.tests.testStarter, since we can't reuse the test case class
# without running the test cases,

class ZopeStarterTestCase(test_logger.LoggingTestBase):

    schema = None
    TEMPNAME = None
    zope_conf = None

    def setUp(self):
        test_logger.LoggingTestBase.setUp(self)
        self.TEMPNAME = mkdtemp()
        if self.schema is None:
            from Zope2.Startup.tests.testStarter import getSchema
            ZopeStarterTestCase.schema = getSchema()

    def tearDown(self):
        rmtree(self.TEMPNAME)
        test_logger.LoggingTestBase.tearDown(self)
        if self.zope_conf is not None:
            del self.zope_conf.servers # should release servers

    def load_config_text(self, text):
        # We have to create a directory of our own since the existence
        # of the directory is checked.  This handles this in a
        # platform-independent way.
        schema = self.schema
        sio = cStringIO.StringIO(
            text.replace("<<INSTANCE_HOME>>", self.TEMPNAME))
        conf, self.handler = ZConfig.loadConfigFile(schema, sio)
        self.assertEqual(conf.instancehome, self.TEMPNAME)
        return conf

    def get_starter(self, conf):
        import Zope2.Startup
        starter = Zope2.Startup.get_starter()
        starter.setConfiguration(conf)
        return starter

    def load_config_text_and_start_servers(self, zope_conf_text):
        from App.config import setConfiguration
        self.zope_conf = self.load_config_text(zope_conf_text)
        starter = self.get_starter(self.zope_conf)
        setConfiguration(self.zope_conf)
    # do the job the 'handler' would have done (call prepare)
        for server in self.zope_conf.servers:
            server.prepare('', None, 'Zope2', {}, None)
        
        starter.setupServers()
        return self.zope_conf

    def test_zserver_views_configuration(self):
        from ZServer.HTTPServer import zhttp_handler
        from Products.ZServerViews.handler import ZServerViewHandler
        from Products.ZServerViews.tests.common import current_thread_id_zserver_view
        import Products.ZServerViews
        zope_conf_text = """
            instancehome <<INSTANCE_HOME>>
            <http-server>
                address 18092
                fast-listen off
            </http-server>
            <ftp-server>
               address 18093
            </ftp-server>

            <product-config zserver-views>
                id-view /foo Products.ZServerViews.tests.common.current_thread_id_zserver_view
            </product-config>
        """
        conf = self.load_config_text_and_start_servers(zope_conf_text)
        import ZServer
        # We should have two configured servers:
        zhttp_server, zftp_server = conf.servers
        self.assertEqual(zhttp_server.__class__,
                         ZServer.HTTPServer.zhttp_server)
        self.assertEqual(zftp_server.__class__,
                         ZServer.FTPServer)
        # At this point, Only the standard zhttp_handler should be present
        # in the HTTP server:
        self.assertEqual([handler.__class__ for handler in zhttp_server.handlers],
                         [zhttp_handler])
        original_handler, = zhttp_server.handlers
        # Now we call the product initialization for Products.ZServerViews.
        # We don't need to pass a real context.
        Products.ZServerViews.initialize(None)
        # This configures the extra handler into the http_server,
        # inserting it first.
        self.assertEqual(
            [handler.__class__ for handler in zhttp_server.handlers],
            [ZServerViewHandler, zhttp_handler],)
        self.assertEqual(zhttp_server.handlers[-1], original_handler,)
        # Our installed handler should have the configuration we passed:
        zserver_handlers = zhttp_server.handlers
        handler, _ = zserver_handlers
        self.assertTrue(handler._match_uri('/foo'))
        env = dict(PATH_INFO='/foo', SCRIPT_NAME='')
        view = handler.get_view(env)
        self.assertEqual(view, current_thread_id_zserver_view)
        self.assertEqual(env, dict(PATH_INFO='', SCRIPT_NAME='/foo'))
        self.assertFalse(handler._match_uri('/bar'))
        self.assertFalse(handler._match_uri('/'))
        # The configuration should be idempotent. I.e. calling it again
        # should result in the same state.
        Products.ZServerViews.initialize(None)
        self.assertEqual(zserver_handlers, zhttp_server.handlers)
        # Emptying the configuration should remove our special handler
        del conf.product_config['zserver-views']
        Products.ZServerViews.initialize(None)
        self.assertEqual([original_handler], zhttp_server.handlers)
        # And there's an API for other products to insert their own keys
        # into our configuration
        Products.ZServerViews.update_configuration(
            {'id-view-2': '/bar Products.ZServerViews.tests.common.current_thread_id_zserver_view'})
        env = dict(PATH_INFO='/bar', SCRIPT_NAME='')
        handler = zhttp_server.handlers[0]
        view = handler.get_view(env)
        self.assertEqual(view, current_thread_id_zserver_view)
        self.assertEqual(env, dict(PATH_INFO='', SCRIPT_NAME='/bar'))
        # the old url for the view no longer works because we had cleared it
        # earlier.
        self.assertTrue(handler._match_uri('/bar'))
        self.assertFalse(handler._match_uri('/foo'))
        # But the update is additive:
        Products.ZServerViews.update_configuration(
            {'id-view-3': '/foo Products.ZServerViews.tests.common.current_thread_id_zserver_view'})
        handler = zhttp_server.handlers[0]
        self.assertTrue(handler._match_uri('/bar'))
        self.assertTrue(handler._match_uri('/foo'))
        self.assertFalse(handler._match_uri('/baz'))


Products/ZServerViews/base.py
##############################################################################
#
# Copyright (c) 2013 Zope Foundation and Contributors.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE
#
##############################################################################
import sys

'''
Base classes for ZServer Views
'''

class ViewError(Exception):

    def __init__(self, status, message, headers=()):
        self.status = status
        self.headers = headers
        self.message = message or u''
        Exception.__init__(self, status + '\n'+ message)

class TextView(object):
    """Decorator for a ZServer view callable to render a small text snippet
    
    The text snippet must be unicode, but the view doesn't have to care about
    start_response(), only return the text
    """

    def __init__(self, func):
        self.func = func

    def __call__(self, environment, start_response):
        headers = [('Content-Type', 'text/plain; charset=utf-8'),]
        try:
            result = self.func(environment)
        except ViewError:
            # python 3 compat: get exception from sys.exc_info()
            # instead of directly
            _c, error, _tb = sys.exc_info()
            status = error.status
            headers.extend(error.headers)
            result = error.message
            del _c, error, _tb
        else:
            status = '200 OK'
        result = result.encode('utf-8')
        length = str(len(result))
        headers.append(('Content-Length', length))
        start_response(status, headers)
        return [result]


Products/ZServerViews/__init__.py
##############################################################################
#
# Copyright (c) 2013 Zope Foundation and Contributors.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE
#
##############################################################################

GRAFT_SERVER_CLASS_SET = set('''
ZServer.HTTPServer.zhttp_server
'''.strip().splitlines())
# TODO: perhaps allow configuration of the set above through zope.conf.
# In any case, it can be safely updated by other modules at import time, since
# it's only used at "Product initialization" time.

CONFIG_KEY = 'zserver-views'

_zserver_view_handler = None

def initialize(context):
    from App.config import getConfiguration
    zope_conf = getConfiguration()
    configure_server_handler(zope_conf)

def remove_handler(server_list):
    global _zserver_view_handler
    if _zserver_view_handler is not None:
        for server in server_list:
            server.remove_handler(_zserver_view_handler)
    _zserver_view_handler = None

def add_handler(handler, server_list):
    global _zserver_view_handler
    _zserver_view_handler = handler
    for server in server_list:
        server.install_handler(_zserver_view_handler)

def configure_server_handler(zope_conf):
    from Products.ZServerViews.handler import ZServerViewHandler
    product_config = get_product_config(zope_conf)
    server_list = get_server_to_graft_list(zope_conf)
    if not server_list:
        return
    if product_config:
        handler_config = get_handler_configuration(product_config)
        if _zserver_view_handler is None:
            handler = ZServerViewHandler(handler_config)
            add_handler(handler, server_list)
        else:
            _zserver_view_handler.update_configuration(handler_config)
    else:
        # configuration is empty, remove handler
        remove_handler(server_list)

def get_product_config(zope_conf):
    return zope_conf.product_config.get(CONFIG_KEY)

def update_configuration(new_conf):
    from App.config import getConfiguration
    zope_conf = getConfiguration()
    product_config = zope_conf.product_config.setdefault(CONFIG_KEY, {})
    product_config.update(new_conf)
    configure_server_handler(zope_conf)

def get_handler_configuration(product_config):
    from Zope2.Startup.datatypes import importable_name
    handler_config = {}
    for entry in product_config.values():
        path, dotted_name = entry.split()
        handler_config[path] = importable_name(dotted_name)
    return handler_config

def get_class_dotted_name(instance):
    cls = instance.__class__
    return cls.__module__ + '.' + cls.__name__

def get_server_to_graft_list(zope_conf):
    result = [server for server in getattr(zope_conf, 'servers', ())
              if get_class_dotted_name(server) in GRAFT_SERVER_CLASS_SET]
    return result


Products/ZServerViews/handler.py
##############################################################################
#
# Copyright (c) 2013 Zope Foundation and Contributors.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE
#
##############################################################################

'''
Handler classes for ZServer Views 
'''
from ZServer.HTTPServer import CONTENT_LENGTH
from ZServer.HTTPServer import zhttp_handler
from ZServer.medusa.http_server import get_header
from ZServer import DebugLogger
from ZServer.HTTPResponse import ChannelPipe as BaseChannelPipe

from logging import getLogger

log = getLogger(__name__)

class ChannelPipe(BaseChannelPipe):

    def start_response(self, status, headers, exc_info=None):
        header_map = dict((header.lower(), value) for header, value in headers)
        # close the connection if requested, or if we have no content length
        if header_map.get('connection', '') =='close':
            self._close = 1
        if (not header_map.get('content-length', None) and
            header_map.get('transfer-encoding', None) != 'chunked'):
            # missing lenght and not chunked encoding: non HTTP/1.1 compliant
            self._close = 1
            self._request.version = '1.0'
        return BaseChannelPipe.start_response(self, status, headers, exc_info)

class ZServerViewHandler(zhttp_handler):
    '''
    Handles connections and serves views directly from the ZServer thread
    '''

    def __init__(self, config):
        zhttp_handler.__init__(self, __name__, uri_base='/')
        self.update_configuration(config)

    def update_configuration(self, config):
        self.config = config

    def match(self, request):
        # NOTE: request.split_url() is more like the deprecated urlparse than
        # urlsplit so we don't use it here
        uri = request.uri
        return self._match_uri(uri)

    def _match_uri(self, uri):
        # We don't need the full urlsplit functionality
        path = uri.split('?', 1)[0]
        return path in self.config

    def get_view(self, env):
        #from ZServer.HTTPResponse import make_response
        #from ZPublisher.HTTPRequest import HTTPRequest
        #request = env['ZServer.medusa.http_server.http_request']
        #stdin = env['stdin']
        #zresponse=make_response(request,env)
        #zresponse._http_connection = 'close'
        #zrequest=HTTPRequest(stdin, env, zresponse)
        path = env['PATH_INFO']
        view = self.config[path]
        env['SCRIPT_NAME'] += path
        env['PATH_INFO'] = ''
        return view

    def continue_request(self, sin, request):
        "continue handling request now that we have the stdin"

        s=get_header(CONTENT_LENGTH, request.header)
        if s:
            s=int(s)
        else:
            s=0
        DebugLogger.log('I', id(request), s)
        env=self.get_environment(request)
        path = request.uri.split('?', 1)[0]
        env['PATH_INFO'] += path
        # why not do the above in self.get_environment(request)?
        env['stdin'] = sin
        env['ZServer.medusa.http_server.http_request'] = request
        try:
            channel_pipe = ChannelPipe(request)
            view = self.get_view(env)
            # mini WSGI server:
            result = view(env, channel_pipe.start_response)
            for part in result:
                # XXX: This needs improvement. Basically, we block ZServer
                # iterating through the answer. If it's long, The ZServer will
                # block while the channel gobbles up all the result in memory.
                # Only after all the result has been absorbed will we return
                # control to the medusa thread to actually serve the view.
                # Instead, we should channel.push() a producer for the answer
                # that iterates the result whenever the medusa thread is ready
                # to push more content.
                #
                # For short content, like debug views, this is ok, but if a
                # view tried, for example, to stream a huge file, this would
                # be bad.
                channel_pipe.write(part)
            channel_pipe.close()
            request.channel.current_request=None
        except:
            log.exception("Failure calling view for %r with env:\n%s\n" %
                          (path, env))
            raise
